### PR TITLE
Support sequence pair for some functions in PretrainedTransformerTokenizer

### PR DESCRIPTION
**Pull request by ZhaofengWu**
`intra_word_tokenize` and `_determine_num_special_tokens_added`

### CODE REVIEW

File: allennlp/data/tokenizers/pretrained_transformer_tokenizer.py [hunk 1]
```diff
@@ -201,50 +228,61 @@ def intra_word_tokenize(self, tokens: List[str]) -> Tuple[List[Token], List[Tupl
...
 
-        return wp_tokens, offsets
-
-    def _determine_num_special_tokens_added(self) -> Tuple[int, int]:
+    def _determine_num_special_tokens_added(self) -> Tuple[int, int, int]:
         """
-        Determines the number of tokens `tokenizer` adds to a sequence (currently doesn't
-        consider sequence pairs) in the start & end.
+        Determines the number of tokens `tokenizer` adds to a sequence or sequence pair
+        in the start & middele & end.
```
Comment (brendan-ai2) [inline allennlp/data/tokenizers/pretrained_transformer_tokenizer.py]: Nit: s/start & middele & end./start, middle and end./

File: allennlp/data/tokenizers/pretrained_transformer_tokenizer.py [hunk 2]
```diff
@@ -182,17 +183,43 @@ def tokenize(self, text: str) -> List[Token]:
         """
         return self._tokenize(text)
 
-    def intra_word_tokenize(self, tokens: List[str]) -> Tuple[List[Token], List[Tuple[int, int]]]:
+    def intra_word_tokenize(self, tokens_a: List[str], tokens_b: Optional[List[str]] = None) -> Any:
```
Comment (brendan-ai2) [inline allennlp/data/tokenizers/pretrained_transformer_tokenizer.py]: Honestly, I'd suggest just making two separate functions `intra_word_tokenize` and `intra_word_tokenize_pair`. The calling code has to be different anyway to account for the different return type, so we might as well have two functions with much simpler type signatures.

File: allennlp/data/tokenizers/pretrained_transformer_tokenizer.py [hunk 3]
```diff
@@ -201,50 +228,61 @@ def intra_word_tokenize(self, tokens: List[str]) -> Tuple[List[Token], List[Tupl
...
         """
-        Determines the number of tokens `tokenizer` adds to a sequence (currently doesn't
-        consider sequence pairs) in the start & end.
+        Determines the number of tokens `tokenizer` adds to a sequence or sequence pair
+        in the start & middele & end.
 
         # Returns
 
-        The number of tokens (`int`) that are inserted in the start & end of a sequence.
+        The number of tokens (`int`) that are inserted in the start & middle & end of a sequence.
```
Comment (brendan-ai2) [inline allennlp/data/tokenizers/pretrained_transformer_tokenizer.py]: And here

File: allennlp/tests/data/tokenizers/pretrained_transformer_tokenizer_test.py [hunk 4]
```diff
@@ -307,9 +307,75 @@ def test_token_idx_sentence_pairs(self):
...
+            (1, 2),
+            (3, 3),
+            (4, 6),
+            (7, 8),
+        ]
+        expected_offsets_b = [
+            (10, 10),
+            (11, 12),
+        ]
+        tokens, offsets_a, offsets_b = tokenizer.intra_word_tokenize_sentence_pair(
```
Comment (brendan-ai2) [inline allennlp/tests/data/tokenizers/pretrained_transformer_tokenizer_test.py]: It's a bit cleaner to just have a separate test case for this function.

Comment (ZhaofengWu) [reply allennlp/tests/data/tokenizers/pretrained_transformer_tokenizer_test.py]: That's what I thought as well but I decided to follow the convention in that test file (https://github.com/allenai/allennlp/blob/4e6c06d8aef916e1f8428dbdba41b9f416353bc9/allennlp/tests/data/tokenizers/pretrained_transformer_tokenizer_test.py#L43).

Comment (brendan-ai2) [reply allennlp/tests/data/tokenizers/pretrained_transformer_tokenizer_test.py]: Good point! Local consistency is useful even when the surrounding code is a bit suboptimal.

**Pull request merged**